### PR TITLE
dissalow symlinks in local storages that point outside the datadir

### PR DESCRIPTION
**Pull request by icewind1991**
For now it's always blocked, will look into adding a mount option seperatly

cc @PVince81

### CODE REVIEW

File: lib/private/Files/Storage/Local.php [hunk 1]
```diff
@@ -337,10 +348,27 @@ public function hasUpdated($path, $time) {
...
 		$fullPath = $this->datadir . $path;
-		return $fullPath;
+		if ($this->allowSymlinks || $path === '') {
+			return $fullPath;
+		}
+		$pathToResolve = $fullPath;
+		$realPath = realpath($pathToResolve);
+		while ($realPath === false) { // for non existing files check the parent directory
+			$pathToResolve = dirname($pathToResolve);
+			$realPath = realpath($pathToResolve);
```
Comment (PVince81) [inline lib/private/Files/Storage/Local.php]: in which cases is this loop necessary ? broken symlinks ?

**Comment by PVince81** [comment]
@icewind1991 from now on please mention @owncloud/filesystem for reviews in this part

**Comment by PVince81** [comment]
- [ ] BUG: `occ files:scan --all` aborts when finding symlink with a bogus error message "Home storage for user admin not writable"

**Comment by PVince81** [comment]
- [ ] BUG: "following symlink not allowed" exception is causing Webdav 500. Please make this a 403

**Comment by PVince81** [comment]
Not sure if a bug, but the disallowed symlink doesn't appear in Webdav's PROPFIND results.
This means it's invisible but a user can still perform operations, like `mkdir symlink`, trying to overwrite it somehow. Still gives a 500 like above.

File: lib/private/Files/Storage/Local.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/Files/Storage/Local.php]: non existing files folders, it checks the first existing parent folder for symlinks

**Comment by PVince81** [comment]
Bugs were found, setting back to "Developing"

**Comment by icewind1991** [comment]
@PVince81 all fixed

**Comment by PVince81** [comment]
@icewind1991 thanks, it works now.

Can you make it so that the invalid symlink still appears in the list ?
Currently it's invisible and it could cause UX trouble or other bugs when trying to perform operations on an invisible but forbidden file.

**Comment by icewind1991** [comment]
It's possible but will take more work since it will make the blocking of symlinks conditional.

I also don't think it's an important enough usecase

**Comment by PVince81** [comment]
Hmm, or it will need a different implementation. Basically let it always be displayed, but when trying to scan or access it, block it with 403 like the firewall does.

I'm ok with merging this as is as it's a corner case, but please let me know what you think about the above idea.

**Comment by icewind1991** [comment]
For it to be displayed id needs to allow scanning

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
Needs second review @owncloud/filesystem @mmattel

**Comment by PVince81** [comment]
@nickvergessen @georgehrke

**Comment by PVince81** [comment]
@guruz @dragotin

**Comment by guruz** [comment]
Looks ok, haven't tested it.. 👍

**Comment by PVince81** [comment]
@icewind1991 can you rebase to kick CI ? Seems even sqlite died :skull:

**Comment by icewind1991** [comment]
jenkins has been poked

**Comment by nickvergessen** [comment]
integration tests fail

**Comment by icewind1991** [comment]
Forgot to update the tests when I changed the exception, should be fine now

**Pull request merged**